upcloud/client/client.go

```diff
@@ -63,6 +68,14 @@ func (c *Client) GetTimeout() time.Duration {
 	return c.httpClient.Timeout
 }
 
+// WithCustomAPI returns a clone of the client with overridden baseUrl and version
+func (c *Client) WithCustomAPI(baseUrl, version string) *Client {
```

Comment:
Minor question: Is there a reason for doing it this way and not just exposing the variables, having Set-methods (like `SetTimeout` that's already there) or allowing them to be used like other configuration - eg. in the initialization methods? 

Adding more permutations to the init methods isn't ideal in any way but tbh I'm not sure why the extra layer of abstraction is a good idea either and this adds something extra to grok / a potential (but probably unlikely) footgun with the cloning (while keeping the same reference to http.Client)?

Reply:
We can do this with Set-methods obviously. This was kind of a precursor for possible upcoming context support where mutating the client state is certainly undesired. We have basically two options there: either introduce new duplicate methods which will lead to potentially long method names `CancelManagedDatabaseConnection` -> `CancelManagedDatabaseConnectionContext` or `CtxCancelManagedDatabaseConnection` while other option being the stdlib http request's `Context` method. Personally I find `srv.WithContext(myCtx).CancelManagedDatabaseConnection(...)` easier to read and use. But I digress.

The api base url override was mainly added while developing this and can be certainly removed from this PR as there's now an envvar based override method for integration testing only. However, we will eventually need this for upctl development.

The With-stuff is purely my personal preference as I'd like an library to be thread-safe from the get go. However, we can change it to mutating method if you prefer it.

Reply:
Another option would be just to enforce context in all of those calls? This would, of course, make it backwards incompatible but would also force the consumer to think about the context. This also seems (to me) to be quite common nowadays in the Go world. 

However, the main concern I have here is consistency. Adding With here would make the client have 3 different ways of configuring it (separate New-methods, Sets, and With-cloning) which I hope we can agree isn't the greatest situation to be in. (As a sidenote, there's also the `New(...opts)`-style somewhere in another repo AFAIR, adding one more style of initialization in our codebase)

Personally it doesn't really matter which style we go for, as long as we keep it consistent. I do agree that this would be a useful feature, too, so not really into dropping it altogether over a question of which style should we go for. That is - let's just make these consistent (either Sets or With or New) and break the backwards compatibility now rather than have the different styles and wanting to break it later? 

Also, still not really sold on the cloning approach, it seems to have plenty of pitfalls that are not obvious and would require essentially manual work - What actually gets copied (eg. is this member an interface or struct or..)? What gets changed when we actually change things in the struct? Why does this `(Foo *s) ChangeSomething()`-method not actually change it in all cases? Why can't you write such methods? Are all the members also safe to be copied / used from multiple threads? Not that I really have a good solution for this either but just spitballing on the challenges ahead..

Reply:
Can say only that I agree with everything. Maybe it would be still beneficial to move this to another PR as it's not right now strictly required?

Personally I'm all for `New(...opts)` style and we can drop the `With` -stuff.

Reply:
SGTM :+1:

Reply:
Not to muddle the conversation further but just to add another datapoint slightly relevant to this - there's also the `NewObject(conf ObjectConfig)`-style in our codebase

Reply:
removed the code from this PR